### ISSUE TITLE
NoGood Forgetting

### ISSUE THREAD

**Issue by AntoniusW** [alpha-asp/Alpha#84]
Alpha uses a CDNL-based solving component which learns new nogoods after running into conflicts. Over time, the number of nogoods becomes very large and many of the learned ones are not helpful to later search. The search almost grinds to a halt due to the massive number of superflous nogoods, which need to be treated during propagation. SAT solvers and other APS solvers therefore implement some form of nogood forgetting in order to speed-up the search and only keep those nogoods that are useful. There is a variety of literature on this topic and Alpha needs some form of nogood forgetting in order to solve the harder instances from the ASP competition.

**Comment by lorenzleutgeb** [comment]
* [Entry point of forgetting in clasp](https://github.com/potassco/clasp/blob/9eafedcfb1ba3048a6e8e8039b9673797763afbe/src/solver.cpp#L1594-L1610)
 * [Entry point for forgetting in SAT4J](https://gitlab.ow2.org/sat4j/sat4j/blob/master/org.sat4j.core/src/main/java/org/sat4j/minisat/core/Solver.java#L1424-1445)
 * [A Bird’s-Eye View of Forgetting in Answer-Set Programming](https://link.springer.com/chapter/10.1007/978-3-319-61660-5_2)
 * [Forgetting in ASP: The Forgotten Properties](https://link.springer.com/chapter/10.1007/978-3-319-48758-8_37)
 * [A Verified SAT Solver Framework with Learn, Forget, Restart, and Incrementality](https://link.springer.com/chapter/10.1007/978-3-319-40229-1_4)
 *

**Comment by AntoniusW** [comment]
Note: the SAT4J link points to variable forgetting, not clause/nogood forgetting. Those are different things.

**Comment by AntoniusW** [comment]
6. [Predicting Learnt Clauses Quality in Modern SAT Solvers](http://ijcai.org/Proceedings/09/Papers/074.pdf) introduces glue clauses and LBD measure as used in the [Glucose](http://www.labri.fr/perso/lsimon/glucose/) SAT solver.

**Comment by lorenzleutgeb** [comment]
> Good idea but keep in mind forgetting leads to incompleteness.

@jblaszczyk No, it does not. Learned nogoods that do not imply anything for the current partial assignment can be forgotten without making the search incomplete.

General note regading this issue: We should not forget that we are grounding lazily, so we might not only forget nogoods, but also datastructures in the grounder. This interferes with grounding heuristics.

**Comment by rtaupe** [comment]
Is this implemented by #197?
